File: android/app/src/main/java/app/notomorrow/feature/fuel/FuelCalendarSheet.kt
package app.notomorrow.feature.fuel

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.requiredSize
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.layout.wrapContentWidth
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.dp
import app.notomorrow.designsystem.Eyebrow
import app.notomorrow.designsystem.Grabber
import app.notomorrow.designsystem.KcalLabel
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtSheet
import app.notomorrow.designsystem.NtShapes
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.TabularText
import app.notomorrow.designsystem.ntPlainClickable
import app.notomorrow.designsystem.pressScale
import app.notomorrow.model.TrainingGoal
import app.notomorrow.util.Fmt
import app.notomorrow.util.NtKeys
import app.notomorrow.util.S
import java.time.LocalDate

/*
 * "History" — the port of `FuelCalendarSheet` (`Features/Fuel/FuelCalendarSheet.swift`): half a
 * year of days as ember cells (brighter = closer to the kcal goal, judged for the current training
 * goal), a legend and three stat tiles. Geometry is the Swift geometry, 1 pt -> 1 dp.
 */

/** `.presentationDetents([.height(450)])`. */
val FuelCalendarSheetHeight = 450.dp

private val CellSize = 24.dp

/** Cell pitch: every tap target is the full 28 × 28 square, the 24 dp fill centred in it. */
private val CellPitch = 28.dp

/** The selected day's ring: 3 dp around the fill, so it overhangs the 28 dp cell by 1 dp a side. */
private val RingSize = CellSize + 6.dp

/**
 * The scrolling grid's inset at each end: twice the ring's overhang (the overhang plus room for
 * the stroke's anti-aliased edge), so the ring on the current week's column — flush against the
 * trailing edge — and on the oldest column is never clipped.
 */
private val GridEdgeInset = RingSize - CellPitch
private val MonthRow = 14.dp
private val MonthGap = 6.dp
private val WeekdayColumn = 22.dp

/**
 * Legend width that is not text: the level-0 swatch, the four-swatch scale (4 × 12 + 3 × 3), the
 * spacer's 12 dp minimum and the five 6 dp gaps between the row's six children.
 */
private val LegendFixedWidth = 12.dp + 57.dp + 12.dp + 6.dp * 5

/**
 * Tapping a day calls [onSelect]; the owner closes the sheet and moves Fuel there. [kcalByDay] is
 * keyed by [FuelCalendar.dayKey]; days with nothing logged are absent.
 */
@Composable
fun FuelCalendarSheet(
    selectedDay: LocalDate,
    today: LocalDate,
    kcalGoal: Double,
    goal: TrainingGoal,
    kcalByDay: Map<LocalDate, Double>,
    onSelect: (LocalDate) -> Unit,
    onDismiss: () -> Unit,
) {
    val layout = remember(today) { FuelCalendar.layout(today) }
    val stats = remember(kcalByDay, today, kcalGoal, goal) {
        FuelCalendar.stats(kcalByDay, today, kcalGoal, goal)
    }

    NtSheet(
        onDismiss = onDismiss,
        containerColor = NT.Colors.ground,
        height = FuelCalendarSheetHeight,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = NT.Spacing.screenH),
        ) {
            Grabber(Modifier.align(Alignment.CenterHorizontally))

            Row(
                modifier = Modifier.fillMaxWidth().padding(top = 10.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                NtText(
                    text = stringResource(S.fuel_calendar_title),
                    style = NT.Fonts.title2,
                    color = NT.Colors.ink,
                    maxLines = 1,
                )
                Spacer(Modifier.weight(1f))
                // `.buttonStyle(.plain)` — tappable, no press feedback.
                Box(
                    modifier = Modifier
                        .height(NT.Size.control)
                        .ntPlainClickable(role = Role.Button, onClick = onDismiss),
                    contentAlignment = Alignment.Center,
                ) {
                    NtText(
                        text = stringResource(S.common_done),
                        style = NT.Fonts.subheadline,
                        color = NT.Colors.ink2,
                        maxLines = 1,
                    )
                }
            }

            // History is scored against today's goal (no goal history is kept), so say which one.
            ShrinkingText(
                text = stringResource(
                    S.fuel_calendar_goalNote,
                    stringResource(NtKeys.goal(goal)),
                    Fmt.kcal(kcalGoal),
                ),
                style = NT.Fonts.footnote,
                color = NT.Colors.ink2,
                minScale = 0.8f,
            )

            FuelCalendarGrid(
                layout = layout,
                selectedDay = selectedDay,
                kcalGoal = kcalGoal,
                goal = goal,
                kcalByDay = kcalByDay,
                onSelect = onSelect,
                modifier = Modifier.padding(top = 18.dp),
            )
            FuelCalendarLegend(Modifier.padding(top = 10.dp))
            FuelCalendarStats(stats, Modifier.padding(top = 18.dp))
        }
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Grid
// ─────────────────────────────────────────────────────────────────────────────

@Composable
private fun FuelCalendarGrid(
    layout: FuelCalendar.Layout,
    selectedDay: LocalDate,
    kcalGoal: Double,
    goal: TrainingGoal,
    kcalByDay: Map<LocalDate, Double>,
    onSelect: (LocalDate) -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.Top,
    ) {
        Column(Modifier.clearAndSetSemantics {}) {
            Spacer(Modifier.height(MonthRow + MonthGap))
            for (isoWeekday in 1..7) {
                Box(
                    modifier = Modifier.size(width = WeekdayColumn, height = CellPitch),
                    contentAlignment = Alignment.CenterStart,
                ) {
                    NtText(
                        text = stringResource(NtKeys.weekday(isoWeekday)),
                        style = NT.Fonts.caption,
                        color = NT.Colors.ink3,
                        maxLines = 1,
                    )
                }
            }
        }

        BoxWithConstraints(
            Modifier
                .weight(1f)
                .height(MonthRow + MonthGap + CellPitch * 7),
        ) {
            val count = layout.columns.size
            val visible = ((maxWidth - GridEdgeInset * 2) / CellPitch).toInt().coerceAtLeast(1)
            val first = FuelCalendar.firstVisibleColumn(layout.columnOf(selectedDay), visible, count)
            // Opens on the current week at the right edge (`defaultScrollAnchor(.trailing)`): asking
            // for the last column lets the list settle flush against its end. A selected day that
            // would be off screen there is centred instead.
            val listState = rememberLazyListState(
                initialFirstVisibleItemIndex = if (first >= count - visible) count - 1 else first,
            )
            LazyRow(
                state = listState,
                modifier = Modifier.fillMaxWidth(),
                contentPadding = PaddingValues(horizontal = GridEdgeInset),
            ) {
                items(count = count, key = { it }) { c ->
                    Column(Modifier.width(CellPitch)) {
                        MonthLabelSlot(layout.monthLabels.firstOrNull { it.column == c })
                        Spacer(Modifier.height(MonthGap))
                        for (day in layout.columns[c]) {
                            CalendarCell(
                                day = day,
                                today = layout.today,
                                selectedDay = selectedDay,
                                kcal = day?.let { kcalByDay[it] },
                                kcalGoal = kcalGoal,
                                goal = goal,
                                onSelect = onSelect,
                            )
                        }
                    }
                }
            }
        }
    }
}

/** Labels are at least two columns apart, so a name wider than one column may run over its neighbour's empty slot. */
@Composable
private fun MonthLabelSlot(label: FuelCalendar.MonthLabel?) {
    Box(Modifier.size(width = CellPitch, height = MonthRow)) {
        if (label != null) {
            NtText(
                text = Fmt.monthShort(label.month),
                modifier = Modifier
                    .wrapContentWidth(Alignment.Start, unbounded = true)
                    .clearAndSetSemantics {},
                style = NT.Fonts.caption,
                color = NT.Colors.ink2,
                maxLines = 1,
            )
        }
    }
}

@Composable
private fun CalendarCell(
    day: LocalDate?,
    today: LocalDate,
    selectedDay: LocalDate,
    kcal: Double?,
    kcalGoal: Double,
    goal: TrainingGoal,
    onSelect: (LocalDate) -> Unit,
) {
    if (day == null) {
        // A future day this week: not drawn, not tappable, not read.
        Spacer(Modifier.size(CellPitch))
        return
    }
    val isToday = day == today
    val isSelected = day == selectedDay
    val level = FuelCalendar.level(
        kcalEaten = kcal ?: 0.0,
        kcalGoal = kcalGoal,
        goal = goal,
        hasEntries = kcal != null,
        isToday = isToday,
    )
    val label = cellLabel(day, kcal, kcalGoal, isToday)
    // TalkBack reads "Double-tap to <action>", so this is the action phrase ("otworzyć ten dzień"),
    // not the iOS hint sentence ("Otwiera ten dzień").
    val action = stringResource(S.fuel_calendar_cell_action)
    Box(
        modifier = Modifier
            .size(CellPitch)
            .pressScale(role = Role.Button, onClickLabel = action) { onSelect(day) }
            .semantics {
                contentDescription = label
                selected = isSelected
            },
        contentAlignment = Alignment.Center,
    ) {
        Box(
            Modifier
                .size(CellSize)
                .background(NT.Colors.heat[level], NtShapes.rounded(6.dp)),
        )
        if (isToday) {
            Box(
                Modifier
                    .size(5.dp)
                    .background(if (level >= 3) NT.Colors.ground else NT.Colors.ink, CircleShape),
            )
        }
        if (isSelected) {
            Box(
                Modifier
                    .requiredSize(RingSize)
                    .border(2.dp, NT.Colors.ink, NtShapes.rounded(9.dp)),
            )
        }
    }
}

/** "Monday, 21 September: 2 140 kcal, 93% of goal" ("… so far …" for today, "… nothing logged" when empty). */
@Composable
private fun cellLabel(day: LocalDate, kcal: Double?, kcalGoal: Double, isToday: Boolean): String {
    val date = Fmt.longDay(day)
    if (kcal == null) return stringResource(S.fuel_calendar_cell_empty, date)
    val percent = if (kcalGoal > 0) Fmt.percent(kcal / kcalGoal) else "–"
    return stringResource(
        if (isToday) S.fuel_calendar_cell_today else S.fuel_calendar_cell,
        date,
        Fmt.kcal(kcal),
        percent,
    )
}

// ─────────────────────────────────────────────────────────────────────────────
// Legend & stats
// ─────────────────────────────────────────────────────────────────────────────

/**
 * `[■0] Not logged · · · Off target [■1■2■3■4] On target`. Swatch 0 means "nothing logged", not
 * "far off", so it sits apart from the off → on target scale. One line; the text shrinks to 0.8
 * before it truncates (`.lineLimit(1).minimumScaleFactor(0.8)`), and it is read as one element.
 */
@Composable
private fun FuelCalendarLegend(modifier: Modifier = Modifier) {
    val none = stringResource(S.fuel_calendar_legend_none)
    val off = stringResource(S.fuel_calendar_legend_off)
    val on = stringResource(S.fuel_calendar_onTarget)
    val measurer = rememberTextMeasurer()
    val density = LocalDensity.current
    BoxWithConstraints(modifier.fillMaxWidth()) {
        val available = constraints.maxWidth
        val style = remember(none, off, on, available, density) {
            val caption = NT.Fonts.caption
            val text = listOf(none, off, on).sumOf {
                measurer.measure(it, caption, maxLines = 1, softWrap = false).size.width
            }
            val room = available - with(density) { LegendFixedWidth.roundToPx() }
            if (text <= 0 || room >= text) caption
            else caption.copy(fontSize = caption.fontSize * (room.toFloat() / text).coerceAtLeast(0.8f))
        }
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .semantics(mergeDescendants = true) {},
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            LegendSwatch(0)
            NtText(none, style = style, color = NT.Colors.ink2, maxLines = 1)
            Spacer(Modifier.weight(1f).widthIn(min = 12.dp))
            NtText(off, style = style, color = NT.Colors.ink2, maxLines = 1)
            Row(horizontalArrangement = Arrangement.spacedBy(3.dp)) {
                for (level in 1..4) LegendSwatch(level)
            }
            NtText(on, style = style, color = NT.Colors.ink2, maxLines = 1)
        }
    }
}

@Composable
private fun LegendSwatch(level: Int) {
    Box(
        Modifier
            .size(12.dp)
            .background(NT.Colors.heat[level], NtShapes.rounded(3.dp)),
    )
}

/**
 * 7-day avg · 30-day avg · on target. The labels may wrap to two lines ("ZGODNIE Z CELEM"); every
 * tile then takes the tallest height and the values stay on one line at the bottom.
 */
@Composable
private fun FuelCalendarStats(stats: FuelCalendar.Stats, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(IntrinsicSize.Min),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        CalendarStatTile(stringResource(S.fuel_calendar_avg7), Modifier.weight(1f)) {
            KcalValue(stats.avg7)
        }
        CalendarStatTile(stringResource(S.fuel_calendar_avg30), Modifier.weight(1f)) {
            KcalValue(stats.avg30)
        }
        CalendarStatTile(stringResource(S.fuel_calendar_onTarget), Modifier.weight(1f)) {
            TabularText(
                text = stringResource(S.fuel_calendar_ofDays, stats.onTarget30, FuelCalendar.STATS_WINDOW),
                style = NT.Fonts.headline,
                color = NT.Colors.ink,
            )
        }
    }
}

@Composable
private fun KcalValue(kcal: Double?) {
    if (kcal != null) {
        KcalLabel(kcal = kcal, numberStyle = NT.Fonts.headline)
    } else {
        TabularText(text = "–", style = NT.Fonts.headline, color = NT.Colors.ink3)
    }
}

/** `StatTile` chrome with any value view underneath a label of up to two lines. */
@Composable
private fun CalendarStatTile(
    label: String,
    modifier: Modifier = Modifier,
    value: @Composable () -> Unit,
) {
    Column(
        modifier = modifier
            .fillMaxHeight()
            .background(NT.Colors.surface, NtShapes.tile)
            .padding(horizontal = 14.dp, vertical = 12.dp)
            .semantics(mergeDescendants = true) {},
        horizontalAlignment = Alignment.Start,
    ) {
        Eyebrow(label, maxLines = 2)
        Spacer(Modifier.height(4.dp))
        Spacer(Modifier.weight(1f))
        value()
    }
}

File: NoTomorrow/Features/Fuel/FuelCalendarSheet.swift
import SwiftUI
import SwiftData

/// "History": half a year of days as ember cells (brighter = closer to the kcal goal, judged for the current training
/// goal), a legend and three stat tiles. Tapping a day opens it in Fuel; the owner closes the sheet.
/// Loads its own data when opened (`FuelCalendar.kcalByDay`); nothing runs while it is closed.
struct FuelCalendarSheet: View {
    let selectedDay: Date
    let kcalGoal: Double
    let goal: TrainingGoal
    var onSelect: (Date) -> Void

    @Environment(\.modelContext) private var modelContext
    @Environment(\.dismiss) private var dismiss
    @State private var kcalByDay: [Date: Double] = [:]

    private let today: Date
    private let layout: FuelCalendar.Layout
    /// The Fuel model's calendar: the day a cell stands for is the day `FuelModel.go(to:)` opens.
    private let calendar: Calendar

    static let height: CGFloat = 450
    private static let cell: CGFloat = 24
    /// Cell pitch: every tap target is the full 28 × 28 square, the 24 pt fill centred in it.
    private static let pitch: CGFloat = 28
    private static let monthRow: CGFloat = 14
    private static let monthGap: CGFloat = 6
    private static let weekdayColumn: CGFloat = 22

    init(selectedDay: Date, kcalGoal: Double, goal: TrainingGoal, calendar: Calendar = .autoupdatingCurrent,
         onSelect: @escaping (Date) -> Void) {
        let today = calendar.startOfDay(for: .now)
        self.calendar = calendar
        self.selectedDay = FuelCalendar.dayKey(selectedDay, calendar: calendar)
        self.kcalGoal = kcalGoal
        self.goal = goal
        self.onSelect = onSelect
        self.today = today
        self.layout = FuelCalendar.layout(today: today, calendar: calendar)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Grabber().frame(maxWidth: .infinity)
            HStack {
                Text("fuel.calendar.title").font(NT.Fonts.title2).foregroundStyle(NT.Colors.ink)
                Spacer()
                Button { dismiss() } label: {
                    Text("common.done").font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2)
                        .frame(height: NT.Size.control)
                }
                .buttonStyle(.plain)
            }
            .padding(.top, 10)

            // History is scored against today's goal (no goal history is kept), so say which one.
            Text(goalNote)
                .font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2)
                .lineLimit(1).minimumScaleFactor(0.8)

            grid.padding(.top, 18)
            legend.padding(.top, 10)
            statTiles.padding(.top, 18)
            Spacer(minLength: 8)
        }
        .padding(.horizontal, NT.Spacing.screenH)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
        .ntScreenBackground()
        .presentationDetents([.height(Self.height)])
        .presentationDragIndicator(.hidden)
        .presentationBackground(NT.Colors.ground)
        .task { kcalByDay = FuelCalendar.kcalByDay(from: layout.start, in: modelContext, calendar: calendar) }
    }

    private var goalNote: String {
        FuelText.format("fuel.calendar.goalNote",
                        goal.localizedName,
                        Fmt.kcal(kcalGoal))
    }

    // MARK: - Grid

    private var grid: some View {
        HStack(alignment: .top, spacing: 6) {
            // Fixed width: a flexible spacer here would let the HStack give this column half the row.
            VStack(alignment: .leading, spacing: 0) {
                Color.clear.frame(width: Self.weekdayColumn, height: Self.monthRow + Self.monthGap)
                ForEach(1...7, id: \.self) { isoWeekday in
                    Text(LocalizedStringKey(AttendanceService.labelKey(isoWeekday: isoWeekday)))
                        .font(NT.Fonts.caption).foregroundStyle(NT.Colors.ink3)
                        .frame(width: Self.weekdayColumn, height: Self.pitch, alignment: .leading)
                }
            }
            .fixedSize(horizontal: true, vertical: false)
            .accessibilityHidden(true)

            GeometryReader { geo in
                ScrollViewReader { proxy in
                    ScrollView(.horizontal) {
                        HStack(spacing: 0) {
                            ForEach(layout.columns.indices, id: \.self) { c in
                                column(c).id(c)
                            }
                        }
                    }
                    .scrollIndicators(.hidden)
                    .defaultScrollAnchor(.trailing)
                    .onAppear { scrollToSelection(proxy, width: geo.size.width) }
                }
            }
            .frame(maxWidth: .infinity)
            .frame(height: Self.monthRow + Self.monthGap + 7 * Self.pitch)
        }
        .accessibilityElement(children: .contain)
    }

    /// Opens on the current week; a selected day that would be off screen there is centred instead.
    private func scrollToSelection(_ proxy: ScrollViewProxy, width: CGFloat) {
        let count = layout.columns.count
        let visible = max(1, Int(width / Self.pitch))
        let first = FuelCalendar.firstVisibleColumn(selected: layout.column(of: selectedDay, calendar: calendar),
                                                    visible: visible, count: count)
        guard first < max(0, count - visible) else { return }   // the trailing anchor already shows it
        proxy.scrollTo(first, anchor: .leading)
    }

    private func column(_ c: Int) -> some View {
        VStack(spacing: 0) {
            monthLabel(c).frame(width: Self.pitch, height: Self.monthRow, alignment: .leading)
            Color.clear.frame(width: Self.pitch, height: Self.monthGap)
            ForEach(0..<7, id: \.self) { r in
                cell(layout.columns[c][r])
            }
        }
    }

    /// Labels are at least two columns apart, so a name wider than one column may run over its neighbour's empty slot.
    @ViewBuilder
    private func monthLabel(_ c: Int) -> some View {
        if let label = layout.monthLabels.first(where: { $0.column == c }) {
            Text(Fmt.monthShort(label.month))
                .font(NT.Fonts.caption).foregroundStyle(NT.Colors.ink2)
                .fixedSize()
                .accessibilityHidden(true)
        } else {
            Color.clear
        }
    }

    @ViewBuilder
    private func cell(_ day: Date?) -> some View {
        if let day {
            let kcal = kcalByDay[day]
            let isToday = day == today
            let isSelected = day == selectedDay
            let level = FuelCalendar.level(kcalEaten: kcal ?? 0, kcalGoal: kcalGoal, goal: goal,
                                           hasEntries: kcal != nil, isToday: isToday)
            Button { onSelect(day) } label: {
                ZStack {
                    RoundedRectangle(cornerRadius: 6, style: .continuous)
                        .fill(NT.Colors.heat[level])
                        .frame(width: Self.cell, height: Self.cell)
                    if isToday {
                        Circle().fill(level >= 3 ? NT.Colors.ground : NT.Colors.ink).frame(width: 5, height: 5)
                    }
                    if isSelected {
                        RoundedRectangle(cornerRadius: 9, style: .continuous)
                            .strokeBorder(NT.Colors.ink, lineWidth: 2)
                            .frame(width: Self.cell + 6, height: Self.cell + 6)
                    }
                }
                .frame(width: Self.pitch, height: Self.pitch)
                .contentShape(Rectangle())
            }
            .buttonStyle(PressScale())
            .accessibilityLabel(Text(cellLabel(day, kcal: kcal, isToday: isToday)))
            .accessibilityHint(Text("fuel.calendar.cell.hint"))
            .accessibilityAddTraits(isSelected ? .isSelected : [])
        } else {
            Color.clear.frame(width: Self.pitch, height: Self.pitch).accessibilityHidden(true)
        }
    }

    /// "Monday, 21 September: 2 140 kcal, 93% of goal" ("… so far …" for today, "… nothing logged" when empty).
    private func cellLabel(_ day: Date, kcal: Double?, isToday: Bool) -> String {
        let date = Fmt.longDay(day)
        guard let kcal else { return FuelText.format("fuel.calendar.cell.empty", date) }
        let percent = kcalGoal > 0 ? Fmt.percent(kcal / kcalGoal) : "–"
        return FuelText.format(isToday ? "fuel.calendar.cell.today" : "fuel.calendar.cell", date, Fmt.kcal(kcal), percent)
    }

    // MARK: - Legend & stats

    /// Swatch 0 means "nothing logged", not "far off", so it sits apart from the off → on target scale.
    private var legend: some View {
        HStack(spacing: 6) {
            swatch(0)
            Text("fuel.calendar.legend.none")
            Spacer(minLength: 12)
            Text("fuel.calendar.legend.off")
            HStack(spacing: 3) {
                ForEach(1...4, id: \.self) { swatch($0) }
            }
            Text("fuel.calendar.onTarget")
        }
        .font(NT.Fonts.caption)
        .foregroundStyle(NT.Colors.ink2)
        .lineLimit(1)
        .minimumScaleFactor(0.8)
        .accessibilityElement(children: .combine)
    }

    private func swatch(_ level: Int) -> some View {
        RoundedRectangle(cornerRadius: 3, style: .continuous)
            .fill(NT.Colors.heat[level])
            .frame(width: 12, height: 12)
    }

    private var statTiles: some View {
        let stats = FuelCalendar.stats(kcalByDay: kcalByDay, today: today, kcalGoal: kcalGoal, goal: goal,
                                       calendar: calendar)
        return HStack(alignment: .top, spacing: 10) {
            statTile("fuel.calendar.avg7") { kcalValue(stats.avg7) }
            statTile("fuel.calendar.avg30") { kcalValue(stats.avg30) }
            statTile("fuel.calendar.onTarget") {
                Text(FuelText.format("fuel.calendar.ofDays", stats.onTarget30, FuelCalendar.statsWindow))
                    .font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink).tabular().lineLimit(1)
            }
        }
        .fixedSize(horizontal: false, vertical: true)
    }

    @ViewBuilder
    private func kcalValue(_ kcal: Double?) -> some View {
        if let kcal {
            KcalLabel(kcal: kcal, numberFont: NT.Fonts.headline)
        } else {
            Text(verbatim: "–").font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink3)
        }
    }

    /// `StatTile` chrome with any value view. Labels may wrap to two lines ("ZGODNIE Z CELEM"); every tile then takes
    /// the tallest height and the values stay on one baseline at the bottom.
    private func statTile<Value: View>(_ label: LocalizedStringKey, @ViewBuilder value: () -> Value) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(label).eyebrow().lineLimit(2).fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
            value()
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 12)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
        .background(NT.Colors.surface, in: RoundedRectangle(cornerRadius: NT.Radius.tile, style: .continuous))
        .accessibilityElement(children: .combine)
    }
}
